File: Backtracking/Word Search.cpp
class Solution {
private: 
    bool solve(int i, int j, int ind, int m, int n, string& word, vector<vector<char>>& board, vector<vector<int>>& vis) {
        if(ind == word.size()) return true;

        if(i < 0 || j < 0 || i >= m || j >= n || vis[i][j] == true || board[i][j] != word[ind]) return false;

        vis[i][j] = true;

        if(solve(i-1, j, ind+1, m, n, word, board, vis)) return true;
        if(solve(i+1, j, ind+1, m, n, word, board, vis)) return true;
        if(solve(i, j+1, ind+1, m, n, word, board, vis)) return true;
        if(solve(i, j-1, ind+1, m, n, word, board, vis)) return true;

        vis[i][j] = false;
        return false;
    }
    
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();

        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(board[i][j] == word[0]) {
                    vector<vector<int>> vis (m, vector<int> (n, 0));
                    if(solve(i, j, 0, m, n, word, board, vis)) return true;
                }
            }
        }
        return false;
    }
};

File: Backtracking/Word Search.rs
use stdcollectionsHashSet;

impl Solution {
    pub fn exist(board VecVecchar, mut word String) - bool {
        word = word.chars().rev().collectString();
        for i in 0..board.len() {
            for j in 0..board[0].len() {
                let mut visited = HashSetnew();
                if board[i][j] == word.chars().next().unwrap() && Selfdfs(i, j, 0, &board, &word, &mut visited) {
                    return true;
                }
            }
        }
        false
    }

    fn dfs(i usize, j usize, curr usize, board &VecVecchar, word &String, visited &mut HashSet(usize, usize)) - bool {
        if curr == word.len() {
            return true;
        }
        
        if i  board.len() && j  board[0].len() && !visited.contains(&(i, j)) && curr  word.len() && board[i][j] == word.chars().nth(curr).unwrap() {
            visited.insert((i, j));

            if Selfdfs(i+1, j, curr+1, board, word, visited) 
                Selfdfs(i-1, j, curr+1, board, word, visited) 
                Selfdfs(i, j+1, curr+1, board, word, visited) 
                Selfdfs(i, j-1, curr+1, board, word, visited) {
                    return true;
            }

            visited.remove(&(i, j));
        }

        false
    }
}
